src/sandstorm/run-bundle.c++

```diff
@@ -1432,6 +1426,25 @@ private:
       tmpfsUidOpts = ",uid=0,gid=0";
     } else {
       unshareUidNamespaceOnce();
```

Comment:
I think this normally _won't_ create a user namespace because we already unshared the user namespace in the parent process -- hence the "once".

What we probably need to do here is change `unshareUidNamespaceOnce()` to unshare every time _if_ `getuid()` returns non-zero.

There's a further complexity with regard to mounting `/proc`. In the logic below, we either mount proc if we're in a PID namespace, or we bind-mount the original proc if not. However, if we're in a PID namespace, it is owned by the parent userns, on which we've already lost capabilities. So this doesn't work anymore. I guess we may have to nest the PID namespaces as well, so we can create a new one here so that it's mountable? Ugh.

Reply:
I think what ends up happening is:

* `sandstorm start` does not create a userns here, as you suggest.
* `sandstorm continue` *does* create a userns, since there is no prior call to this in that codepath. This is where it counts, after the exec.

I don't know why the fork seems to be necessary.

This raises another concern: If we're spawning a new userns on *each* exec, is it possible this is going to grow without bound, adding another level of nesting after each update? This probably wouldn't grow that quickly, but I could forsee some future point in time where we hit a limit and everybody's server crashes...

Reply:
The `unsharedUidNamespace` flag is actually passed through `sandstorm continue` via the `--userns` flag, so in fact a user NS is never created here during either `sandstorm start` or `sandstorm continue`. (I think the reason this line exists is for the call site from `sandstorm mongo`.)

My guess is that this change works for the initial startup because in that case there's no `exec()` where the capabilities are dropped, but it probably won't work after the first update.

I think we probably need to create both a nested userns and a nested pidns. And I think to do that without creating yet another process in the hierarchy, we might want to use `clone()` instead of `exec()` to create the server monitor, so that we can pass `CLONE_NEWUSER` and `CLONE_NEWPID` at that time, without affecting the state of the parent process.